### PR TITLE
Adding more control for defined flows

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
In the WaterTAP costing package, some variables and mutable parameters, while settable, do not have any impact on the underlying Pyomo model. A fix is coming in IDAES, (see the discussion on IDAES/idaes-pse#1014), but in the meantime WaterTAP can side-step this issue.

## Changes proposed in this PR:
- Register defined flow costs on the WaterTAPCosting package through `add_defined_flow`.
- Add additional safeguards to the `defined_flow` dictionary so values cannot be set by `__setitem__` directly.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Review by TimBartholomew — APPROVED** [review]
LGTM

**Comment by TimBartholomew** [comment]
@lbibl can you review/approve this?

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/tests/test_dye_desalination_withRO.py [hunk 1]
```diff
@@ -128,8 +128,8 @@ def test_costing(self, system_frame):
         assert_optimal_termination(results)
 
         # check costing
-        assert pytest.approx(0.653284, rel=1e-3) == value(m.fs.LCOW)
-        assert pytest.approx(0.270094, rel=1e-3) == value(m.fs.LCOT)
+        assert pytest.approx(0.690498, rel=1e-3) == value(m.fs.LCOW)
+        assert pytest.approx(0.284743, rel=1e-3) == value(m.fs.LCOT)
```
Comment (lbibl) [inline watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/tests/test_dye_desalination_withRO.py:132]: I couldn't see issues with other changes.  Only what is causing the value changes here?

**Review by lbibl — APPROVED** [review]
If the two value changes are correct for sure,  it looks good to me.

File: watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/tests/test_dye_desalination_withRO.py (hunk 1 above)
Comment (bknueven) [reply watertap/examples/flowsheets/case_studies/wastewater_resource_recovery/dye_desalination/tests/test_dye_desalination_withRO.py:132]: The electricity cost for the RO module wasn't getting properly set. So in the old computation $0.07/kWh was used. With the bug fix the correct value of $0.08/kWh is used instead.

**Pull request merged**